### ISSUE TITLE
Be able to set key and secret for a user for riak-cs [JIRA: RCS-213]

### ISSUE THREAD

**Issue by sdebnath** [basho/riak_cs#1074]
In dev environments, having a new key and secret combination generated every time a new user in the cluster is set up leads to unnecessary work that can be easily avoided by providing an option to generate secure keys and secrets with an utility and then having the option to specify the key and secret when registering an user or setting/re-setting it after the fact. This allows developers to get up and running quickly and maintain the same credentials across cluster teardown and refreshes. This is particularly true of folks using Docker to get their riak-cs clusters set up. 

To not break S3 and other API compatibility, Basho could provide an alternate API to enable this feature. This could potentially also be used in PROD by doing credential push from AD type infrastructure.

**Comment by ghost** [comment]
That would indeed be very useful.

Maybe the easiest way to achieve this would be to extend the add users API () by allowing to specify access key and secret key; for this, `accept_json(RD, Ctx=#context{user=undefined})` and `accept_xml(RD, Ctx=#context{user=undefined})` in https://github.com/basho/riak_cs/blob/develop/src/riak_cs_wm_user.erl would have to be modified to optionally accept more JSON/XML parameters, and if they are present, call `riak_cs_user:create_user/4` instead of `riak_cs_user:create_user/2`.

**Comment by HMcKelvey** [comment]
Are you looking for an API that supports AD and maybe OWASP integration?

On Thu, Jun 11, 2015 at 5:07 AM, Felix Fontein notifications@github.com
wrote:

> That would indeed be very useful.
> 
> Maybe the easiest way to achieve this would be to extend the add users API
> () by allowing to specify access key and secret key; for this, accept_json(RD,
> Ctx=#context{user=undefined}) and accept_xml(RD,
> Ctx=#context{user=undefined}) in
> https://github.com/basho/riak_cs/blob/develop/src/riak_cs_wm_user.erl
> would have to be modified to optionally accept more JSON/XML parameters,
> and if they are present, call riak_cs_user:create_user/4 instead of
> riak_cs_user:create_user/2.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/basho/riak_cs/issues/1074#issuecomment-111052498.

## 

Heather McKelvey
VP Engineering
Basho Technologies

**Comment by ghost** [comment]
I'm personally only looking for a very simple JSON (or XML) based API which allows me to create the admin user with a pre-specified Access Key and Secret Key.

**Comment by kuenishi** [comment]
Although it depends on the context what exactly 'dev' means, I usually run following one-liner after Riak, Stanchion and CS process kicked up. `admin-key` and `admin-secret` is sorta preset default key and secret so you can utilise them.

```
$ /usr/lib/riak-cs/erts-5.9.1/bin/erl  -name n@127.0.0.1 -setcookie riak -eval "io:format(\"~p\", [rpc:call('riak-cs@127.0.0.1', riak_cs_user, create_user, [\"name\", \"me@admin.com\", \"admin-key\",  \"admin-secret\"])])." -s init stop
```

**Comment by ghost** [comment]
@kuenishi: that should do the trick for me! Thanks a lot.

It still might be nice to have a more 'official' way to do this, though. Especially if someone (like sdebnath mentions) wants to create a whole set of users with pre-specified keys.

**Comment by sdebnath** [comment]
Yep, its useful to basically manage users in an automated way in production, revoking or granting access as necessary with the help of an external system
